Source/CesiumRuntime/Private/CesiumEncodedMetadataConversions.h

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020-2023 CesiumGS, Inc. and Contributors
+
+#pragma once
+
+enum class ECesiumEncodedMetadataType : uint8;
+struct FCesiumPropertyTablePropertyDescription;
+struct FCesiumPropertyTableProperty;
+struct FCesiumMetadataPropertyDetails;
+struct FCesiumMetadataEncodingDetails;
+
+/**
+ * @brief Gets the best-fitting encoded types and conversion method for a given
+ * metadata type. This determines the best way (if one is possible) to transfer
+ * values of the given type to the GPU, for access in Unreal materials.
+ *
+ * An array size can also be supplied if bIsArray is true on the given value
+ * type. If bIsArray is true, but the given array size is zero, this indicates
+ * the arrays of the property vary in length. Variable-length array properties
+ * are unsupported.
+ *
+ * @param ValueType The given value type.
+ * @param ArraySize The size of the array, if applicable.
+ */
+FCesiumMetadataEncodingDetails CesiumMetadataPropertyDetailsToEncodingDetails(
+    FCesiumMetadataPropertyDetails PropertyDetails);
+
+/**
+ * @brief Gets the number of components associated with the given encoded type.
+ * @param type The encoded metadata type.
+ */
+size_t
+CesiumGetEncodedMetadataTypeComponentCount(ECesiumEncodedMetadataType type);
+
+/**
+ * Any custom encoding behavior, e.g., special encoding of unsupported
+ * properties, can go here. Use the below methods as examples.
+ */
+
+/**
+ * @brief Coerces property values to the type specified by the property
+ * description. The following property types are supported:
+ * - scalars
+ * - vecNs
+ * - booleans
+ * - scalar and boolean arrays (up to the first four elements)
+ *
+ * Additionally, if the property contains strings or string arrays, it will
+ * attempt to parse numbers from each string, then coerce those numbers to the
+ * desired format.
+ */
+struct CesiumEncodedMetadataCoerce {
+  /**
+   * Whether it is possible to apply the encoding method based on the property
+   * description.
+   *
+   * @param description The property table property description.
+   */
+  static bool
+  canEncode(const FCesiumPropertyTablePropertyDescription& description);
+
+  /**
+   * Encodes the data of the property table property into the given texture data
+   * pointer, as the type specified in the property description.
+   *
+   * @param propertyDescription The property table property description.
+   * @param property The property table property itself.
+   * @param pTextureData A pointer to the texture data, which will be filled
+   * during encoding.
+   * @param pixelSize The size of a pixel from the given texture, in bytes.
+   */
+  static void encode(
+      const FCesiumPropertyTablePropertyDescription& propertyDescription,
+      const FCesiumPropertyTableProperty& property,
+      void* pTextureData,
```

Comment:
I wanted this function to encode to the texture directly, without creating an intermediate data structure. But I don't know the best way to do this 😬 please let me know if this is atrocious!

Reply:
Is this the raw pointer you were asking about in our call today? If so, I think the only thing I might suggest improving about it is to be clearer about the required size of the `pTextureData` buffer. So documentation, mostly. But it might be better yet to pass a `gsl::span<std::byte>` rather than a `void*`. Then you can throw if the buffer isn't big enough. That will terminate the process if exceptions are disabled, but that's much better than corrupting memory.